File: src/Partition_List.java
import util.ListNode;
import util.Print;

/**
 * 2014-7-22
 * @author Alvin
 * 
 */

public class Partition_List {

    public ListNode partition(ListNode head, int x) {
    	ListNode small = null;
    	ListNode smallStep = null;
    	ListNode bigStep = null;
    	ListNode big = null;
    	while(head != null){
    		if(head.val < x){
    			if(smallStep == null){
    				small = head;
    				smallStep = small;
    			}else{
    				smallStep.next = head;
    				smallStep  = smallStep.next;
    			}
    		}else{
    			if(bigStep == null){
    				big = head;
    				bigStep = big;
    			}else{
    				bigStep.next = head;
    				bigStep  = bigStep.next;
    			}
    		}
    		
    		head = head.next;
    	}
    	bigStep.next = null;
    	if(small == null){
    		
    		return big;
    	}else{
    		smallStep.next = big;
    		return small;
    	}
    }
	public static void main(String[] args) {
		// 1->4->3->2->5->2
		ListNode n1 = new ListNode(1);
		ListNode n2 = new ListNode(4);
		ListNode n3 = new ListNode(3);
		ListNode n4 = new ListNode(2);
		ListNode n5 = new ListNode(5);
		ListNode n6 = new ListNode(2);
		n1.next = n2;
		n2.next = n3;
		n3.next = n4;
		n4.next = n5;
		n5.next = n6;
		
		
		
		Print.print(new Partition_List().partition(n1, 3));
	}

}

File: src/cpp/partition-list.cpp
#include <iostream>
#include <vector>
using namespace std;
struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *partition(ListNode *head, int x) 
    {
    	ListNode headLeft(0);
    	ListNode headRight(0);
    	ListNode* pLeft = &headLeft;
    	ListNode* pRight = &headRight;
		ListNode* ptr = head;
		while( ptr != NULL )
		{
			if ( ptr->val < x )
			{
				pLeft->next = ptr;
				pLeft = pLeft->next;
			}
			else
			{
				pRight->next = ptr;
				pRight = pRight->next;
			}
			ptr = ptr->next;
		}
		if ( pLeft != NULL )
		{
			pLeft->next = headRight.next;
		}

		if ( pRight != NULL ) pRight->next = NULL;
		if ( pLeft == &headLeft ) return headRight.next;
		else return headLeft.next;
    }
};

void printList(ListNode *node)
{
	char buffer[1024*10];
	memset(buffer,0,sizeof(char) * 1024 * 10);
	while( node != NULL )
	{

		sprintf(buffer,"%s,%d",buffer,node->val);
		node = node->next;
	}
	printf("%s\n", buffer);
}
ListNode* arr2node(int value[],int len)
{
	ListNode head(0);
	ListNode* p = &head;
	for (int i = 0; i < len; ++i)
	{
		ListNode* tmp = new ListNode(value[i]);
		p->next = tmp;
		p = p->next;
	}
	return head.next;
}

int main(int argc, char const *argv[])
{
	int value1[] = {1,4,3,2,5,2};

	ListNode* head1 = arr2node(value1,sizeof(value1)/sizeof(int));



	Solution s;
	ListNode* result = s.partition(head1,1);
	printList(result);

	return 0;
}